Implemented protocol export through Zip::OutputStream [SCI-742]

okriuchykhin opened the pull request:
https://biosistemika.atlassian.net/browse/SCI-742

mz3944 commented on app/controllers/protocols_controller.rb:
```diff
@@ -532,12 +532,60 @@ def import
+          file_name = 'protocol.eln'
+          unless protocol_name.nil?
+            escaped_name = protocol_name.gsub(/[^0-9a-zA-Z-.,_]/i, '_')
+                                        .downcase[0..250]
```
I would replace 250 with Constants::NAME_MAX_LENGTH.

mz3944 commented on app/utilities/protocols_exporter.rb:
```diff
@@ -1,89 +1,290 @@
+    str2 = id.to_s
+    (19 - str2.size).times do
+      str2 = '0' + str2
+    end
```
Simplify to _str2 = '0' * (19 - str2.size) + str2_

mz3944 commented on app/controllers/protocols_controller.rb:
```diff
@@ -532,12 +532,60 @@ def import
+              protocol.steps.order(:id).each do |step|
+                step_guid = get_guid(step.id)
+                step_dir = "#{protocol_dir}/#{step_guid}"
+                if step.assets.count > 0
```
Same as before.

mz3944 commented on app/controllers/protocols_controller.rb:
```diff
@@ -532,12 +532,60 @@ def import
+            ostream.put_next_entry("#{protocol_dir}/eln.xml")
+            ostream.print(generate_protocol_xml(protocol))
+            # Add assets to protocol folder
+            if protocol.steps.count > 0
```
Use next so that the Hound will shut up :P. Basically, if an if statement is at the end, you should use next instead.

okriuchykhin commented:
I fixed turbolinks progress bar and added export functionality to protocol management page. But it seems problematic to make spinner work again as now we simply provide download link and can not check when download starts and finishes.

mz3944 commented on app/assets/javascripts/protocols/index.js:
```diff
@@ -502,6 +501,20 @@ function updateButtons() {
+    }
+    params = encodeURI(params);
+    exportLink.href = $("[data-action='export']").data('export-url') + params;
+    exportLink.click();
```
Can't you handle this with _window.location.href_? This seems too hack-ish...

Ducz0r commented:
Once you two fix all things in this PR (so it's ready for merge), can you rebase the PR onto `master`?

mz3944 commented:
What's the status on this @okriuchykhin? I think we said we can ignore the fact that there is no spinner, right? If so, you can merge the stuff.

Ducz0r commented:
\* Rebase, and then merge the stuff :smile:.
